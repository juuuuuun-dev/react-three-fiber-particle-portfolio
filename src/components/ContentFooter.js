import React from 'react';
import useStore from '../contexts/store';

const ContentFooter = () => {
  const nowYear = new Date().getFullYear();
  const domain = useStore(state => state.domain);
  const startYear = useStore(state => state.copyrightStartYear);
  const currnetYear = startYear !== nowYear ? `-${nowYear}` : '';
  return (
    <>
      <div className='contents__footer'>
        &copy; {startYear}
        {currnetYear} {domain}
      </div>
    </>
  );
};

export default ContentFooter;
