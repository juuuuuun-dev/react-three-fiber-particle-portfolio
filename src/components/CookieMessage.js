import React from "react"
import useStore from '../contexts/store';
import { get, set } from 'local-storage';

const CookieMessage = () => {
  const actions = useStore(state => state.actions);
  const okCookieKey = "ok_cookie";
  const cookiePath = '/cookies';
  const okCookie = get(okCookieKey);
  const [ok, setOK] = React.useState(okCookie);


  const handleLink = (e) => {
    e.preventDefault();
    actions.toggleContents(cookiePath);
  }

  const handleOK = () => {
    set(okCookieKey, 1)
    setOK(1)
  }

  return (
    <>
    {!ok &&
      <div className="cookie-message">
        <p className="cookie-message__body">This website use cookies. <a href={`${cookiePath}`} onClick={(e) => handleLink(e)}>Learn more</a>.</p>
        <p className="cookie-message__btn" onClick={handleOK}>OK</p>
      </div>
    }
      
    </>
  )
}

export default CookieMessage