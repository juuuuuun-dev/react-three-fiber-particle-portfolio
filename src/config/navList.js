import img1 from '../assets/images/curly_braces.png';
import img2 from '../assets/images/q.png';
import img3 from '../assets/images/contact.png';

export default [
  {
    title: 'Home',
    src: img1,
    texts: ['I am a', 'web', 'developer',],
    textSize: 4.4,
    lineHeight: 3.0,
    bottomY: -9.5,
    positionX: -12.3,
    lotationY: 0.77,
    description: 'descriptionです'
  },
  {
    title: 'ABOUT',
    src: img2,
    texts: ['ABOUT', 'ME'],
    path: '/about',
  },
  {
    title: 'CONTACT',
    src: img3,
    texts: ['COTACT', 'ME'],
    path: '/contact',
  }
];
