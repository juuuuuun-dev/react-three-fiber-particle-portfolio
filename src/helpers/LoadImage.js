export default function(src) {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.crossOrigin = 'Anonymouse';
    img.onload = () => resolve(img);
    img.onerror = e => reject(e);
    img.src = src;
  });
}
