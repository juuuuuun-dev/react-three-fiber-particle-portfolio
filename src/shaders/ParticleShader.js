import * as THREE from 'three'

const ParticleShader = {
  uniforms: {
    uTime: { type: 'f', value: 0 },
    uCoefficient: { type: 'f', value:0 },
    uMousePosition: { type: 'v2', value:  new THREE.Vector2(0.5, 0.5) },
    lightPosition: { type: 'v3', value: new THREE.Vector3(10, 50, 10) },
    lightColor: { type: 'c', value: new THREE.Color("#43465a") },
    ambientLightColor: {value: null},
    lightProbe: {value: null},
    directionalLights: {value: null},
    spotLights: {value: null},
    rectAreaLights: {value: null},
    pointLights: {value: null},
    hemisphereLights: {value: null},
    directionalShadowMap: {value: null},
    directionalShadowMatrix: {value: null},
    spotShadowMap: {value: null},
    spotShadowMatrix: {value: null},
    pointShadowMap: {value: null},
    pointShadowMatrix: {value: null},

  },
  vertexShader: `
    attribute float aAlpha;
    attribute vec3  aColor;
    attribute float aTime;
    attribute vec3 aTarget;
    
    uniform float uTime;
    uniform vec4 uMousePosition;
    uniform float uCoefficient;
    varying float vAlpha;
    varying vec3 vColor;
    varying vec3 vNormal;
    varying vec3 mPosition;
    varying vec3 mvPosition;
    
    highp float random(vec2 co)
    {
        highp float a = 12.9898;
        highp float b = 78.233;
        highp float c = 43758.5453;
        highp float dt= dot(co.xy ,vec2(a,b));
        highp float sn= mod(dt,3.14);
        return fract(sin(sn) * c);
    }
    
    void main(){
    
        float curTime = mod(uTime + aTime, 1.3);
        // float rate = curTime/3.; 
        // todo この数字をattributeに
        float rate = curTime/10.; 
        vec3 pos;
        
        if(rate < 1.){
            pos = mix(position, aTarget * uMousePosition.x * -uCoefficient, rate);
        }else{
            float fract = rate + 1.0;
            pos = mix(position, aTarget * 1.0, 1.0 + fract * fract);
        }
    
        vec4 mvPosition = modelViewMatrix * vec4(pos, 1.0);
        mPosition = mvPosition.xyz;
        if(aAlpha < 0.7) vAlpha =  clamp((1.0 - rate)/1.0, 0., 1.0);
        // else             vAlpha =  clamp((1.0 - rate)/0.5, 0., 1.0);
        vColor = aColor;
        vNormal = normalMatrix * normal;
        gl_PointSize = 0.5; //(size * scale) * (1000.0 / length(mvPosition.xyz));
        gl_Position = projectionMatrix * mvPosition;
        
    }
  `,

  fragmentShader: `
    varying float vAlpha;
    varying vec3 vColor;
    varying vec3 vNormal;
    uniform vec3 lightPosition;
    uniform vec3 lightColor;
    uniform vec3 ambientColor;
    uniform vec3 mPosition;
    uniform vec3 mvPosition;
    void main() {
      vec4 viewLightPosition = viewMatrix * vec4( lightPosition, 0.0);
        vec3 pColor = abs(normalize(mPosition.xyz));
        vec3 N = normalize(vNormal);
        vec3 L = normalize(viewLightPosition.xyz);
        float dotNL = dot(N, L);
        vec3 diffuse = pColor * lightColor  * max(dotNL, 0.0);
        vec3 ambient = pColor * ambientColor;
        // gl_FragColor = vec4(vColor, vAlpha );
        gl_FragColor = vec4(vColor * vColor, vAlpha );
    }
  `,
}

export default ParticleShader;