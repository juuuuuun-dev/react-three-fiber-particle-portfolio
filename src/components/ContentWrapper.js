import React from 'react';
import useStore from '../contexts/store';
import { useTranslation } from 'react-i18next';
import { useSprings, animated } from 'react-spring';
import ContentFooter from './ContentFooter';
import variables from '../scss/_variables.scss';

export default function ContentWrapper({ contentTitle, children }) {
  const lang = useStore(state => state.lang);
  const languages = useStore(state => state.languages);
  const refs = React.useRef(languages.map((_, index) => index));
  const actions = useStore(state => state.actions);
  const contentData = actions.getContentData(contentTitle);
  const { i18n } = useTranslation();
  const textColor = variables.textColor;
  const activeColor = variables.activeColor;

  const springFunc = index => {
    let color = textColor;
    if (lang && lang === languages[index].id) {
      color = activeColor;
    }
    return {
      ref: refs[index],
      color
    };
  };
  const [langSprings, setLangSprings] = useSprings(languages.length, index =>
    springFunc(index)
  );
  setLangSprings(index => springFunc(index));

  React.useEffect(() => {
    i18n.changeLanguage(lang);
  }, [lang, i18n]);
  return (
    <>
      <div className='scroll-area'>
        <div className='contents__inner'>
          <h2 className='contents__title'>
            {contentData.title}{' '}
            <span className='lang-change'>
              {langSprings.map((item, index) => {
                return (
                  <animated.span
                    key={index}
                    style={{ ...item }}
                    onClick={() =>
                      actions.setLang(languages[index].id, contentData.path)
                    }
                  >
                    {languages[index].name}
                  </animated.span>
                );
              })}
            </span>
          </h2>

          <p
            className='back-btn'
            onClick={() => actions.toggleContents(contentData.path)}
          >
            BACK
          </p>
          <div>{children}</div>
          <ContentFooter />
        </div>
      </div>
    </>
  );
}
