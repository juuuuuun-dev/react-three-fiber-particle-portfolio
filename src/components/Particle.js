import React, { useRef, useMemo, useEffect } from 'react';
import { useFrame, useThree } from 'react-three-fiber';
import ImageList from '../config/ImageList';
import LoadImage from '../helpers/LoadImage';
import * as THREE from 'three/src/Three';
import ParticleShader from '../shaders/ParticleShader'
import useStore from '../contexts/store'


/**
 * @todo loading
 * @todo contents表示時 text color
 * @todo mouse 固定
 */

export default function Particle() {
  const actions = useStore(state => state.actions);
  const navListLength = useStore(state => state.navListLength);
  const ua = useStore(state => state.ua);
  let listIndex = 0;
  let time = 0;
  let theta = 0;
  const clock = new THREE.Clock();
  console.log('memo')
  let MAX = useMemo(() => {
    
    if (ua === 'sp') {
      // @todo android firefoxは半分ぐらいでいいかも
      return 10000;
    }
    return 24000;
  }, [ ua ]);
  const bufferAttribute = {
    positions: [],
    endPositions: [],
    alphas: [],
    colors: [],
    times: [],
  }
  const attributes = [];
  const mouse = new THREE.Vector2();
  mouse.x = -0.5;
  mouse.y = 0;
  const mousePos = { x: -1.3, y: 0, px:0, py:0, tx:0, ty:0 };
  const geometryRef = useRef();
  const materialRef = useRef();

  useEffect(() => {
    const f = async () => {
      const imagePositions = await createImagePositions();
      const attribute = await createAttributes(attributes, imagePositions, MAX);
      await (() => {
        return new Promise(resolve => {
          for (let i = 0; navListLength > i; i++) {
            bufferAttribute.positions[i] = new THREE.BufferAttribute(attribute.positions[i], 3);
            bufferAttribute.endPositions[i] = new THREE.BufferAttribute(attribute.endPositions[i], 3);
            bufferAttribute.times[i] = new THREE.BufferAttribute(attribute.times[i], 1);
            bufferAttribute.alphas[i] = new THREE.BufferAttribute(attribute.alphas[i], 1);
            bufferAttribute.colors[i] = new THREE.BufferAttribute(attribute.colors[i], 3);
          }
          geometryRef.current.setAttribute('position', bufferAttribute.positions[listIndex]);
          geometryRef.current.setAttribute('position', bufferAttribute.positions[listIndex]);
          geometryRef.current.setAttribute('aTarget', bufferAttribute.endPositions[listIndex]);
          geometryRef.current.setAttribute('aAlpha', bufferAttribute.alphas[listIndex]);
          geometryRef.current.setAttribute('aTime', bufferAttribute.times[listIndex]);
          geometryRef.current.setAttribute('aColor', bufferAttribute.colors[listIndex]);
          actions.setReady(true);
          actions.setLoading(false);
          actions.setRouter();
          resolve();
        });
      })();
      // ¥
    };
    f();
  }, [ bufferAttribute, actions, listIndex, attributes, MAX, navListLength ]);
  
  const { camera } = useThree();
  let coefficient = 12.0; // first coefficient
  const targetCoefficient = 1.0;
  const mouseTargetCoefficient = -0.5; // z
  useFrame(() => {
    coefficient += (targetCoefficient - coefficient) * .1;
    mouse.x += (mouseTargetCoefficient - mouse.x) * .1;
    let delta = clock.getDelta();
    time += delta;
    theta += (mouse.x / 4 - theta) / 10;
    camera.position.z = 15 * Math.cos(theta);
    camera.position.x = 20 * Math.sin(theta);
    camera.position.y = 80 * theta + 100;
    camera.lookAt(new THREE.Vector3());
    materialRef.current.uniforms.uTime.value = time;
    materialRef.current.uniforms.uCoefficient.value = coefficient;
    materialRef.current.needsUpdate = true;
    materialRef.current.uniforms.uMousePosition.value = mousePos;
  });
  // listIndexをstate(store)にしてしまうと再描画が起こるのでcollbackで処理
  const scrollCollback = (index) => {
    listIndex = index;
    coefficient = 15.6;
    mouse.x = 2;
    if (geometryRef.current) {
      geometryRef.current.setAttribute('position', bufferAttribute.positions[listIndex]);
      geometryRef.current.setAttribute('aTarget', bufferAttribute.endPositions[listIndex]);
      geometryRef.current.setAttribute('aTime', bufferAttribute.times[listIndex]);
      geometryRef.current.setAttribute('aAlpha', bufferAttribute.alphas[listIndex]);
      geometryRef.current.setAttribute('aColor', bufferAttribute.colors[listIndex]);
    }
  };
  actions.setScrollCollbacks(scrollCollback);

  return (
    <>
    <points>
      <bufferGeometry
        attach="geometry"
        ref={geometryRef}
      >
        <bufferAttribute
          needsUpdate={true}
          attachObject={['attributes', 'position']}
          count={MAX}
          array={new Float32Array(MAX * 3)}
          itemSize={3}
        />
      </bufferGeometry>
      <shaderMaterial
        ref={materialRef}
        attach="material"
        name="material"
        lights={true}
        args={[ParticleShader]}
      />
    </points>
    </>
  )
}



/**
 * createImagePositions
 */
function createImagePositions() {
  return new Promise(async (resolve) => {
    const canvas = document.createElement("canvas");
    let positions = [];
    const listLen = ImageList.length;
    for (let i = 0; listLen > i; i++) {
      const image = await LoadImage(ImageList[i].src);
      positions[i] = setImagePosition(image, canvas, i);
    }
    resolve(positions);
  });
}

function setImagePosition(image, canvas, index) {
  let pos = [];
  const scale = 9;
  canvas.width = image.width;
  canvas.height = image.height;
  const ctx = canvas.getContext('2d');
  ctx.drawImage(image, 0, 0);
  const data = ctx.getImageData(0, 0, image.width, image.height);
  const imageData = data.data;
  
  for (let y = 0; y < image.height; y++) {
    for (let x = 0; x < image.width; x++ ) {
      const num = 4 * (image.width * y + x) + 3;
      const alpha = imageData[num];
      if (alpha !== 0) {
        const rRate = x / image.width * 1.1;
        // const rRate = x / image.width * 1.5;
        // const gRate = y / image.height * .9;
        const color = new THREE.Color();
        // color.setRGB(rRate, gRate, 1);
        // color.setRGB(rRate, .8, .85);
        // color.setRGB(.2, .7, 1); #暫定
        color.setRGB(.1, .65, .9);
        const data = {
          x: (x - image.width / 2) / scale,
          y: (y - image.height / 2) / scale,
          alpha: alpha / 255,
          color,
        }
        pos.push(data);
      }
    }
  }
  return pos;
}

/**
 * createAttributes
 */
function createAttributes(attributes, imagePositions, MAX) {
  return new Promise((resolve) => {
    const count = 3;
    // let attr = attributes;
    let attr = {
      positions: [],
      endPositions: [],
      alphas: [],
      colors: [],
      times: [],
    };
    let imageLen = imagePositions.length;
    
    for (let i = 0; imageLen > i; i++) {
      attr.positions[i] = new Float32Array(MAX * count);
      attr.endPositions[i] = new Float32Array(MAX * count);
      attr.alphas[i] = new Float32Array(MAX);
      attr.colors[i] = new Float32Array(MAX * count);
      attr.times[i] = new Float32Array(MAX);
      const datas = imagePositions[i];
      const datasLen = datas.length;
      for (let n = 0; n < MAX; n++) {
        const data = datas[parseInt(datasLen * Math.random())];
        attr.positions[i][n * count] = data.x;
        attr.positions[i][n * count + 1] = 0;
        attr.positions[i][n * count + 2] = data.y;

        attr.endPositions[i][n * count] = range(0, 40);
        attr.endPositions[i][n * count + 1] = range(0, 40);
        attr.endPositions[i][n * count + 2] = range(0, 0);

        attr.times[i][n] = count * Math.random();
        attr.alphas[i][n] = Math.random();
        attr.colors[i][n * count] = data.color.r;
        attr.colors[i][n * count + 1] = data.color.g;
        attr.colors[i][n * count + 2] = data.color.b;
      }
    }
    resolve(attr);
  });
}

function range(min, max){
  return min + (max - min) * Math.random();
}