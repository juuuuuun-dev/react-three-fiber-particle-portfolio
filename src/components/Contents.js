import React from "react";
import ContentWrapper from "./ContentWrapper";
import useStore from "../contexts/store";
import { useSprings, animated } from "react-spring";

/**
 * @todo useSprings collback
 */
export default function Contents() {
  const showContent = useStore(state => state.showContent);
  const actions = useStore(state => state.actions);
  const windowHeight = useStore(state => state.windowHeight);
  const navList = actions.getHasPathNavList();
  const refs = React.useRef(navList.map(() => React.createRef()))
  const springFunc = (index) => {
    let top = windowHeight + 'px';
    let opacity = 0;
    if (showContent && showContent === navList[index].path) {
      top = 0 + "px";
      opacity = 1;
      refs.current[index].current.firstChild.scrollTop = 0;
    }
    return {
      ref: refs[index],
      top,
      opacity,
    }
  }
  const [ springs, setSprings ] = useSprings(navList.length, index => (springFunc(index)));
  setSprings(index => (springFunc(index)));

  return (
    <>
    {springs.map((value, index) => {
        return (
          <animated.div
            key={index}
            style={{...value}}
            className="contents"
            ref={refs.current[index]}
          >
            <ContentWrapper contentTitle={navList[index].title}>
              <p className="text-lerge">lets create something
              <br />beary great together!</p>
              <p className="text-lerge">lets create something
              <br />beary great together!</p>
              <p className="text-lerge">lets create something
              <br />beary great together!</p>
              <p className="text-lerge">lets create something
              <br />beary great together!</p>
              <p className="text-lerge">lets create something
              <br />beary great together!</p>
              <p className="text-lerge">lets create something
              <br />beary great together!</p>
              <p className="text-lerge">lets create something
              <br />beary great together!</p>
              <p className="text-lerge">lets create something
              <br />beary great together!</p>
              <p className="text-lerge">lets create something
              <br />beary great together!</p>
              <p className="text-lerge">lets create something
              <br />beary great together!</p>
              <p className="text-lerge">lets create something
              <br />beary great together!</p>
              <p className="text-lerge">lets create something
              <br />beary great together!</p>
              <h3>center</h3>
              <p className="text-lerge">lets create something
              <br />beary great together!</p>
              <p className="text-lerge">lets create something
              <br />beary great together!</p>
              <p className="text-lerge">lets create something
              <br />beary great together!</p>
              <p className="text-lerge">lets create something
              <br />beary great together!</p>
              <p className="text-lerge">lets create something
              <br />beary great together!</p>
              <p className="text-lerge">lets create something
              <br />beary great together!</p>
              <p className="text-lerge">lets create something
              <br />beary great together!</p>
              <p className="text-lerge">lets create something
              <br />beary great together!</p>
              <p className="text-lerge">lets create something
              <br />beary great together!</p>
              <p className="text-lerge">lets create something
              <br />beary great together!</p>
              <h3>end</h3>
            </ContentWrapper>
          </animated.div>
        );
    })}
    </>
  )
}
