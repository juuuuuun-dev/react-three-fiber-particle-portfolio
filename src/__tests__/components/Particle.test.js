import React from 'react';
import { render, cleanup, getByTestId } from '@testing-library/react'
import { Particle, cameraFrame, materialFrame, createImagePositions, setImagePosition, createAttributes } from '../../components/Particle';
import { Canvas } from 'react-three-fiber';
import navList from '../../config/navList'
import * as THREE from 'three';
import ParticleShader from '../../shaders/ParticleShader';

// import img from '../../images/smile.png'
import LoadImage from '../../helpers/LoadImage'
import 'jest-canvas-mock';
import "@testing-library/jest-dom/extend-expect";
// three
import "mutationobserver-shim";

// jest.mock('../../images/smile.png');
// global.Promise = jest.requireActual('promise');
let canvas;
const img = new Image();
img.src = 'https://placekitten.com/100/50';
img.width = 100;
img.height = 50;
beforeEach(() => {
  canvas = document.createElement('canvas');
});
describe("Particle", () => {
  it('renders Particle', async () => {
    const { container } = render(<Canvas><Particle /></Canvas>);
    const COLOR = [1, 1, 1];
    const navListLength = navList.length;
    const imagePositions = createImagePositions(COLOR, navList, navListLength);
    expect(imagePositions).toBeDefined();
    const position = setImagePosition(img, canvas, 1, COLOR);
    expect(position).toBeDefined();
  });
  it('createAttributes', async () => {
    const MAX = 10;
    let positions = [];
    const position = [{
      x: 1,
      y: 1,
      alpha: 1,
      color: { r: 1, g: 1, b: 1 }
    }]
    for (let i = 0; i <= MAX; i++) {
      positions.push(position);
    }
    const attribute = await createAttributes(positions, MAX / 3);
    expect(attribute.positions).toBeDefined();
    expect(attribute.endPositions).toBeDefined();
    expect(attribute.times).toBeDefined();
    expect(attribute.alphas).toBeDefined();
    expect(attribute.colors).toBeDefined();
  });
  it('cameraFrame', () => {
    const camera = new THREE.PerspectiveCamera(45);
    const theta = 1;
    expect(camera.position.x).toBe(0);
    expect(camera.position.y).toBe(0);
    expect(camera.position.z).toBe(0);
    cameraFrame(camera, theta);
    expect(camera.position.x).not.toBe(0);
    expect(camera.position.y).not.toBe(0);
    expect(camera.position.z).not.toBe(0);
  });
  it('materialFrame', () => {
    const materialRef = {
      current: {
        uniforms: ParticleShader.uniforms
      }
    }
    const time = 10;
    const coefficient = 10;
    const mousePos = { x: 1, y: 0 }
    expect(materialRef.current.uniforms).toBe(ParticleShader.uniforms)
    materialFrame(materialRef, time, coefficient, mousePos);
    expect(materialRef.current.uniforms.uTime.value).toBe(time);
    expect(materialRef.current.uniforms.uCoefficient.value).toBe(coefficient);
    expect(materialRef.current.uniforms.uMousePosition.value).toBe(mousePos)
  });
  it('should be callable', () => {
    const ctx = canvas.getContext('2d');
    ctx.drawImage(img, 1, 2);
    expect(ctx.drawImage).toBeCalled();
  });

});
