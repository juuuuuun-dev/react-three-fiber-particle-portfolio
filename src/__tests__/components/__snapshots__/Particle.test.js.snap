// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Particle renders Particle 1`] = `Array []`;
